diff --git a/include/libimobiledevice/debugserver.h b/include/libimobiledevice/debugserver.h
index 19a4f5d..5db2f61 100644
--- a/include/libimobiledevice/debugserver.h
+++ b/include/libimobiledevice/debugserver.h
@@ -39,6 +39,7 @@ typedef enum {
 	DEBUGSERVER_E_MUX_ERROR      = -2,
 	DEBUGSERVER_E_SSL_ERROR      = -3,
 	DEBUGSERVER_E_RESPONSE_ERROR = -4,
+	DEBUGSERVER_E_TIMEOUT        = -5,
 	DEBUGSERVER_E_UNKNOWN_ERROR  = -256
 } debugserver_error_t;
 
diff --git a/src/debugserver.c b/src/debugserver.c
index 1e06233..83ec864 100644
--- a/src/debugserver.c
+++ b/src/debugserver.c
@@ -54,6 +54,8 @@ static debugserver_error_t debugserver_error(service_error_t err)
 			return DEBUGSERVER_E_MUX_ERROR;
 		case SERVICE_E_SSL_ERROR:
 			return DEBUGSERVER_E_SSL_ERROR;
+        case SERVICE_E_TIMEOUT:
+			return DEBUGSERVER_E_TIMEOUT;
 		default:
 			break;
 	}
